### PR TITLE
Add gather operation

### PR DESCRIPTION
**Pull request by naoyam**
Add a new Fusion IR operation, `gather`, that fetches a window of a given shape for each element.

Example:
```
t0: 2D tensor of [N, M]
// gather a 1x3 window for each element with padding of 1 along the second axis
t1 = gather(t0, {1, 3}, {{0, 0}, {1, 1}});
```
Then:
```
t1: [N, M, 1, 3]
t1[i, j, k, l] = The value at the window position of [k, l] for t0[i, j]
```

2D conv of input `inp` with weight `w` can be expressed as:
```
auto inp_tile = gather(inp, {1, 3, 3}, {{0, 0}, {1, 1}, {1, 1}});
auto inp_bc = broadcast(inp_tile, {true, false, false, false, false, false, false});
auto w_bc = broadcast(w, {false, false, true, true, true, false, false});
auto inp_times_w = mul(inp_bc, w_bc);
auto out = sum(inp_times_w, {1, 4, 5, 6});
```

The shape of the weight tensor can be dynamic. See the FusionConv2DDynamic test.

The basic codegen machinery for shift is reused. The support of dynamic window shape requires some non-trivial changes in lower_shift.cpp, but no fundamental change was needed.

Limitations
- While it is possible to specify padding sizes, it is currently only supported when the shape of the input tensor does not change. Supporting, for example, gather with no padding remains to be done.
- The predicate elimination (#959) is not applicable yet.

Update
- Added the `FusionConv2DDynamicChain` test just for a demonstration of fusing two convolutions.

### CODE REVIEW

File: test/cpp/jit/test_gpu_shift.cpp [hunk 1]
```diff
@@ -2292,6 +2332,433 @@ TEST(NVFuserTest, FusionMaxPooling_CUDA) {
...
+
+  at_inp = at_inp.unsqueeze(0); // at::conv2d needs the N axis
+  auto at_out = at::conv2d(at_inp, at_w, {}, 1, 1);
+  at_out = at_out.squeeze(0); // drop the N axis
+
+  testValidate(&fusion, cg_outputs, inputs, {at_out}, __LINE__, __FILE__);
+}
+
+// Mostly the same as the static conv test, but the shape of the weights,
+// 3x3 in this case, is given dynamically
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu_shift.cpp:2505]: Awesome.

File: test/cpp/jit/test_gpu_shift.cpp [hunk 2]
```diff
@@ -2292,6 +2332,433 @@ TEST(NVFuserTest, FusionMaxPooling_CUDA) {
...
+
+  at_inp = at_inp.unsqueeze(0); // at::conv2d needs the N axis
+  auto at_out = at::conv2d(at_inp, at_w, {}, 1, 1);
+  at_out = at_out.squeeze(0); // drop the N axis
+
+  testValidate(&fusion, cg_outputs, inputs, {at_out}, __LINE__, __FILE__);
+}
+
+// 5x5 followed by 3x3
+TEST(NVFuserTest, FusionConv2DDynamicChain_CUDA) {
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu_shift.cpp:2616]: This is really cool.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 3]
```diff
@@ -1179,6 +1179,99 @@ TensorView* shift(TensorView* inp, const std::vector<int>& offsets) {
...
+    const auto inp_axis = inp_dom[i];
+    const auto window_dim = window_shape[i];
+    const auto pad_left = pad_width[i][0];
+    const auto pad_right = pad_width[i][1];
+    TORCH_INTERNAL_ASSERT(inp_axis->start()->isZeroInt());
+    Val* out_axis_dim = nullptr;
+    if (window_dim->isConst() && pad_left->isConst() && pad_right->isConst()) {
+      const int extent_adjustment =
+          -(-window_dim->value().value() + 1 + pad_left->value().value() +
+            pad_right->value().value());
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1245]: Does this work for even sized windows? I think the math may need to change based on even vs odd window sizes?

File: torch/csrc/jit/codegen/cuda/arith.h [hunk 4]
```diff
@@ -284,6 +284,34 @@ TORCH_CUDA_CU_API TensorView* shift(
...
     const std::vector<int>& offsets);
 
+//! Gather a window of nearby elements for each element.
+//!
+//! Each window of size window_shape is stored as a additional
+//! innermost domain, meaning that the number of dimensions of the
+//! output tensor doubles. The pad_width parameter specifies the
+//! padding width of each side of each axis.
+//!
+//! Example:
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/arith.h:294]: So padding is `[axis][side]` ?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 5]
```diff
@@ -350,22 +357,81 @@ kir::Val* getProducerIndexWithHalo(
...
+
+  // take the root index of the corresponding gather axis
+  const auto window_axis = gather_op->gatherAxis(consumer_axis);
+  auto concrete_id = gpu_lower->caIndexMap().getConcreteMappedID(
+      consumer_tv->getRootDomain().at(window_axis));
+  auto ref_concrete_map_it = ref_concrete_map.find(concrete_id);
+  TORCH_INTERNAL_ASSERT(ref_concrete_map_it != ref_concrete_map.end());
+  IterDomain* reference_root_of_gather_axis = ref_concrete_map_it->second;
+
+  auto window_idx =
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:435]: Looks right to me, but could you add a comment with an example to make it easier to understand the indexing you're doing below.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 6]
```diff
@@ -1030,6 +1095,14 @@ std::vector<kir::Val*> Index::getGlobalProducerStridedIndices(
 
     root_ind = getProducerIndexWithHalo(producer_tv, i, root_ind, consumer_tv);
 
+    root_ind = getProducerIndexWithGather(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:1120]: Can both halo and gather be active at the same time?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 7]
```diff
@@ -350,22 +357,81 @@ kir::Val* getProducerIndexWithHalo(
...
   return producer_index;
 }
 
+//! Offset a producer index of a gather expression
+kir::Val* getProducerIndexWithGather(
+    size_t producer_axis,
+    kir::Val* producer_index,
+    const TensorView* producer_tv,
+    const TensorView* consumer_tv,
+    const std::unordered_map<kir::IterDomain*, kir::Val*>& index_map,
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: nit: reference index map

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 8]
```diff
@@ -537,6 +537,101 @@ bool ShiftOp::sameAs(const Statement* other) const {
...
+  for (size_t i = 0; padWidth().size(); ++i) {
+    if (!padWidth()[i][0]->sameAs(other_op->padWidth()[i][0]) ||
+        !padWidth()[i][1]->sameAs(other_op->padWidth()[i][1])) {
+      return false;
+    }
+  }
+  return Expr::sameAs(other);
+}
+
+int GatherOp::gatherAxis(int axis) const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp:629]: NIT: support negative indexing for consistency.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 9]
```diff
@@ -114,6 +148,54 @@ void ShiftPredicateInserter::insert(
...
+    const std::vector<kir::Val*>& indices) {
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  if (gather_expr == nullptr || axis >= gather_expr->windowShape().size() ||
+      gather_expr->windowShape()[axis]->isOneInt()) {
+    return consumer_index;
+  }
+
+  const auto window_axis = gather_expr->gatherAxis(axis);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp]: Same as index compute:
"Looks right to me, but could you add a comment with an example to make it easier to understand the indexing you're doing below."

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 10]
```diff
@@ -166,49 +249,74 @@ kir::Bool* ShiftPredicateInserter::getPredicate(
...
       // "left" means halo at offset zero.
       // shifted accesses when idx >= left_limit. padding if idx <
       // left_limit.
 
       // The elements at the left halo region are just set by the
       // padding value.
-      unsigned left_limit = halo_info.width(0);
+      kir::Val* left_limit = halo_info.width(0);
+      kir::Val* right_limit = makeAddExpr(
+          out_tv->domain()->rootDomain()[i]->extent(), halo_info.width(0));
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:277]: Why isn't this `width(1)`? I thought width had left and right side in it.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 11]
```diff
@@ -166,49 +249,74 @@ kir::Bool* ShiftPredicateInserter::getPredicate(
...
+
+      kir::Val* consumer_index = indices[i];
+      kir::Val* producer_index = nullptr;
+
+      if (shift_expr != nullptr) {
+        producer_index = getShiftProducerIndex(i, consumer_index, shift_expr);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:283]: Can you add some comment to getShiftProducerIndex and getGatherProducerIndex that they're only valid for predicates as it's a global memory view, not local/shared?

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 12]
```diff
@@ -379,13 +550,30 @@ void HaloInfo::propagateRootAxisInfo(
...
+      const auto window_dim =
+          gpu_lower->lowerValue(gather_op->windowShape()[i]);
+      if (window_dim->isOneInt()) {
+        p_info.merge(c_info);
+        continue;
       }
+      const auto& pad_dim = gather_op->padWidth()[i];
+      const auto pad_dim0 = gpu_lower->lowerValue(pad_dim[0])->as<kir::Int>();
+      p_info.merge(0, makeAddExpr(c_info.width(0), pad_dim0));
+      p_info.merge(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:589]: I'm having some trouble following this, can you add a comment with an example to help me understand?

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 13]
```diff
@@ -415,16 +604,16 @@ void HaloInfo::build(TensorDomain* td) {
         " of ",
         td->getRootDomain());
 
-    if (halo_width == 0) {
-      halo_width_map_.insert({root_axis, 0});
+    if (!halo_info.hasHalo()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:626]: Why did you go full symbolic on this part instead of using your `makeAddExpr` and `makeSubExpr`?

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 14]
```diff
@@ -722,11 +905,54 @@ bool extentCompare(
...
+    auto xv = x->value();
+    auto yv = y->value();
+    return xv.has_value() && yv.has_value() && xv.value() <= yv.value();
+  };
+  return extentCompare(*this, id1, id2, cmp);
 }
 
 bool HaloInfo::extentEqual(IterDomain* id1, IterDomain* id2) const {
-  return extentCompare(*this, id1, id2, std::equal_to<unsigned>());
+  std::function<bool(kir::Int*, kir::Int*)> cmp = [&](kir::Int* x,
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:946]: Could you add some comments on assumptions in this comparison? It seems it can't catch all cases, but if both DAGs producing the ints are exactly the same including all inputs then it will detect it is the same.

**Review by csarofeen — APPROVED** [review]
Overall I think this PR looks really good, so approving now. I did add a few comments, but primarily I'm just looking for more comments in the code explaining the arithmetic as it's a little hard to follow.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp:1245]: I believe this particular logic, the extent of the output axis, is fine. I haven't actually tried even sized windows. The problem is it doesn't work if the output extent differs from the input, so each end of the axis must have a different padding width. As far as I see, it's not possible with `at::conv2d`. I'll see if I could come up with a workaround.

File: torch/csrc/jit/codegen/cuda/arith.h (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.h:294]: Yes

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:435]: Done

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp:1245]: Added a 2x2 conv test. See `FusionConv2DStaticEvenSizedWindow_CUDA` (https://github.com/csarofeen/pytorch/blob/gather/test/cpp/jit/test_gpu_shift.cpp#L2765)

The padding is only used for the right end with size 1 so that the resulting axis has the same size as the input axis. `at::conv2d` doesn't seem to have a different padding size at each end, so I used padding 1 and dropped the first element in the aten case.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 7 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: Done

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:1120]: Yes, and that's actually a common case. When an intermediate tensor is gathered, it's expanded with halo to accommodate the neighbor accesses, so the gather producer does have halo, and `getProducerIndexWithHalo` should adjust the producer indexing for the hallo attached to the producer. And since the expression itself is `gather`, `getProducerIndexWithGather` should adjust the index accordingly. 

These parts of indexing are getting disorganized as I add more operations. For producer indexing, the adjustments we need to do are done when:

- when the producer has halo
- when the expression is shift
- when the expression is gather

Previously, both of the halo and shift adjustment are done in `getProducerHaloOffset`. For more clarity, the shift part of the adjustment may be better if it's extracted out of the function.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 8 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp:629]: Done

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 9 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp]: Added comments

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 10 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:277]: Yes, `width(0)` means the left halo width. I added comments to explain what left and right limits correspond to.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 11 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:283]: Well, the same logic could apply to local/shared tensors as well. In fact, they basically have the same underlying logic as for `getProducerIndexWithHalo` and `getProducerIndexWithGather`. Yes, these functions should be refactored to minimize overlap, which I plan to do.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 13 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:626]: Nothing particular. It was just because it needed additional overloaded versions of `makeAddExpr`, so I just didn't bother to use it. 

Expression simplification is necessary for shift validation, e.g., when we want to prove `<symbolic-val-X> + 1` and `<symbolic-val-X> + (0 + 1)` are the same. It may make more sense to make these functions as part of the KIR builder.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 12 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:589]: Added a comment

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 14 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:946]: Yes, it just checks if two `kir::Int` objects have the same expression trees. As stated in the comment in the header, this is a conservative analysis and only returns true when the equality is proven. That is just sufficient for the current use case, but it could be refactored to something more generic if necessary. Added comments.
